main.cpp

```diff
@@ -16,19 +17,56 @@ std::ostream &operator<<(std::ostream &os, std::vector<T> const &a) {
 
 // 请修复这个函数的定义：10 分
 template <class T1, class T2>
-std::vector<T0> operator+(std::vector<T1> const &a, std::vector<T2> const &b) {
+auto operator+(std::vector<T1> const &a, std::vector<T2> const &b) {
     // 请实现列表的逐元素加法！10 分
     // 例如 {1, 2} + {3, 4} = {4, 6}
+    using T = std::decay_t<decltype(a[0]+b[0])>; //I find the output type when given two vectors.
+    std::vector<T> ret;
+    for(int i=0; i<std::min(a.size(), b.size()); i++){
+        ret.push_back(a[i]+b[i]);
+    }
+    return ret;
 }
 
 template <class T1, class T2>
 std::variant<T1, T2> operator+(std::variant<T1, T2> const &a, std::variant<T1, T2> const &b) {
     // 请实现自动匹配容器中具体类型的加法！10 分
+    std::variant<T1, T2> ret;
+    std::visit([&](auto const &aa, auto const &bb){ret = aa + bb;}, a, b);
+    return ret;
+}
+template<class T1, class T2, class T3>
+auto operator+(std::variant<T1, T2> const& a, T3 const& b){
+    /* 
+    因为出现了std::variant跟std::vector进行相加的运算
+    然而在之前的代码中没有出现相应的输入类型
+    因此对这种加法进行overload
+    */
+    std::variant<T1,T2> ret = b;
+    return a + ret;
 }
 
-template <class T1, class T2>
-std::ostream &operator<<(std::ostream &os, std::variant<T1, T2> const &a) {
-    // 请实现自动匹配容器中具体类型的打印！10 分
+template<class T1, class T2, class T3>
+auto operator+(T3 const& a, std::variant<T1, T2> const& b){
+    return b + a;
+}
+
+
+// template <class T1, class T2>
+// std::ostream &operator<<(std::ostream &os, std::variant<T1, T2> const &a) {
+//     std::visit([&](auto const& input){os<<input;}, a);
+//     return os;
+// }
+
+template<class... T>
+std::ostream &operator<<(std::ostream &os, std::variant<T...> &inputs){
```

Comment:
```suggestion
template<class T0, class... T>
std::ostream &operator<<(std::ostream &os, std::variant<T0,T...> const &inputs){
```
保证至少有一个参数，即可用上 const &。

Reply:
```suggestion
template<class... T, std::enable_if_t<sizeof...(T) != 0, int> = 0>
std::ostream &operator<<(std::ostream &os, std::variant<T...> &inputs){
```
也可以这样。